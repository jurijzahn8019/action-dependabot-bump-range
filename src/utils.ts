/* eslint-disable camelcase */
import { SemVer, coerce, diff, ReleaseType } from "semver";
import { promises as fs } from "fs";

const { readFile } = fs;
const matchPattern = /\w+\([\w-]+\):\s+bump\s+\S+\s+from\s+v?(?<from>[\d.]+)\s+to\s+v?(?<to>[\d.]+)/;

export function matchTitle(
  title: string
): { from: SemVer; to: SemVer } | undefined {
  const match = matchPattern.exec(title);
  if (match?.groups && match.groups.from && match.groups.to) {
    const res = {
      from: coerce(match.groups.from),
      to: coerce(match.groups.to),
    };
    return res.from && res.to ? { from: res.from, to: res.to } : undefined;
  }

  return undefined;
}

export function verDiff(from: SemVer, to: SemVer): ReleaseType | null {
  return diff(from, to);
}

export async function getEvent(): Promise<{
  pull_request?: { number?: number };
}> {
  return JSON.parse(
    await readFile(process.env.GITHUB_EVENT_PATH as string, "utf8")
  );
}
