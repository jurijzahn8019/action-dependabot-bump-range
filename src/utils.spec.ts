import { SemVer } from "semver";
import { matchTitle, verDiff } from "./utils";

const titles = [
  "chore(deps-dev): bump @rollup/plugin-commonjs from 11.1.0 to 13.0.0",
  "chore(deps-dev): bump @otto-ec/toolbox from 1.13.0 to 1.13.1",
  "chore(deps-dev): bump @mdi/font from 5.2.45 to 5.3.45",
  "chore(deps-dev): bump jest from 25.4.0 to 26.0.1",
  "chore(deps): bump actions/setup-node from v1 to v2.0.0",
  "chore(deps-dev): foo bar 35.4.0",
];

describe("utils/matchTitle", () => {
  it("Should match title", () => {
    titles.forEach((t) => expect(matchTitle(t)).toMatchSnapshot(t));
  });
});

describe("utils/verDiff", () => {
  it("Should compare versions", () => {
    [
      { from: "0.3.4", to: "0.4.0" },
      { from: "1.3.4", to: "1.3.5" },
      { from: "1.3.4", to: "1.4.5" },
      { from: "1.3.4", to: "2.4.5" },
    ].forEach((v) =>
      expect(verDiff(new SemVer(v.from), new SemVer(v.to))).toMatchSnapshot(
        `${v.from}:${v.to}`
      )
    );
  });
});
