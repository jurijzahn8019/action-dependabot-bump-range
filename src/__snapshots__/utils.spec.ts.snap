// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/matchTitle Should match title: chore(deps): bump actions/setup-node from v1 to v2.0.0 1`] = `
Object {
  "from": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 1,
    "minor": 0,
    "options": Object {},
    "patch": 0,
    "prerelease": Array [],
    "raw": "1.0.0",
    "version": "1.0.0",
  },
  "to": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 2,
    "minor": 0,
    "options": Object {},
    "patch": 0,
    "prerelease": Array [],
    "raw": "2.0.0",
    "version": "2.0.0",
  },
}
`;

exports[`utils/matchTitle Should match title: chore(deps-dev): bump @mdi/font from 5.2.45 to 5.3.45 1`] = `
Object {
  "from": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 5,
    "minor": 2,
    "options": Object {},
    "patch": 45,
    "prerelease": Array [],
    "raw": "5.2.45",
    "version": "5.2.45",
  },
  "to": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 5,
    "minor": 3,
    "options": Object {},
    "patch": 45,
    "prerelease": Array [],
    "raw": "5.3.45",
    "version": "5.3.45",
  },
}
`;

exports[`utils/matchTitle Should match title: chore(deps-dev): bump @otto-ec/toolbox from 1.13.0 to 1.13.1 1`] = `
Object {
  "from": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 1,
    "minor": 13,
    "options": Object {},
    "patch": 0,
    "prerelease": Array [],
    "raw": "1.13.0",
    "version": "1.13.0",
  },
  "to": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 1,
    "minor": 13,
    "options": Object {},
    "patch": 1,
    "prerelease": Array [],
    "raw": "1.13.1",
    "version": "1.13.1",
  },
}
`;

exports[`utils/matchTitle Should match title: chore(deps-dev): bump @rollup/plugin-commonjs from 11.1.0 to 13.0.0 1`] = `
Object {
  "from": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 11,
    "minor": 1,
    "options": Object {},
    "patch": 0,
    "prerelease": Array [],
    "raw": "11.1.0",
    "version": "11.1.0",
  },
  "to": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 13,
    "minor": 0,
    "options": Object {},
    "patch": 0,
    "prerelease": Array [],
    "raw": "13.0.0",
    "version": "13.0.0",
  },
}
`;

exports[`utils/matchTitle Should match title: chore(deps-dev): bump jest from 25.4.0 to 26.0.1 1`] = `
Object {
  "from": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 25,
    "minor": 4,
    "options": Object {},
    "patch": 0,
    "prerelease": Array [],
    "raw": "25.4.0",
    "version": "25.4.0",
  },
  "to": SemVer {
    "build": Array [],
    "includePrerelease": false,
    "loose": false,
    "major": 26,
    "minor": 0,
    "options": Object {},
    "patch": 1,
    "prerelease": Array [],
    "raw": "26.0.1",
    "version": "26.0.1",
  },
}
`;

exports[`utils/matchTitle Should match title: chore(deps-dev): foo bar 35.4.0 1`] = `undefined`;

exports[`utils/verDiff Should compare versions: 0.3.4:0.4.0 1`] = `"minor"`;

exports[`utils/verDiff Should compare versions: 1.3.4:1.3.5 1`] = `"patch"`;

exports[`utils/verDiff Should compare versions: 1.3.4:1.4.5 1`] = `"minor"`;

exports[`utils/verDiff Should compare versions: 1.3.4:2.4.5 1`] = `"major"`;
